integrations/operator/embeddedtbot/bot.go

```diff
@@ -0,0 +1,138 @@
+package embeddedtbot
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/tbot"
+	"github.com/gravitational/teleport/lib/tbot/config"
+)
+
+type EmbeddedBot struct {
+	cfg *config.BotConfig
+
+	errGroup *errgroup.Group
+
+	credential *config.UnstableClientCredentialOutput
+}
+
+func New(botConfig *BotConfig) (*EmbeddedBot, error) {
+	credential := &config.UnstableClientCredentialOutput{}
+
+	cfg := (*config.BotConfig)(botConfig)
+	cfg.Storage = &config.StorageConfig{Destination: &config.DestinationMemory{}}
+	cfg.Outputs = []config.Output{credential}
+
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bot := &EmbeddedBot{
+		cfg:        cfg,
+		credential: credential,
+	}
+
+	return bot, nil
+
+}
+
+func (b *EmbeddedBot) Preflight(ctx context.Context) (*proto.PingResponse, error) {
+	b.cfg.Oneshot = true
+	bot := tbot.New(b.cfg, log.StandardLogger())
+	err := bot.Run(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	teleportClient, err := b.buildClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pong, err := teleportClient.Ping(ctx)
+	return &pong, trace.Wrap(err)
+}
+
+// start the bot and immediately returns. Once must follow if the bot is still healthy by running
+func (b *EmbeddedBot) start(ctx context.Context) {
+	b.cfg.Oneshot = false
+	bot := tbot.New(b.cfg, log.StandardLogger())
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return bot.Run(egCtx)
+	})
+	b.errGroup = eg
+}
+
+// Start is a lie, the bot is already started. DO NOT CALL Start if you want to run the bot, call StartAndWaitForClient.
+// We mimick a legitimate Start() behaviour by returning if the bot exists and propagating context cancellation.
+func (b *EmbeddedBot) Start(ctx context.Context) error {
+	if b.errGroup == nil {
+		return errors.New("b.Start() is called but StartAndWaitForClient() has not been invoked yet, aborting")
+	}
+
+	// Start a goroutine that waits for the errorGroup and sends back the error
+	errCh := make(chan error)
+	go func() {
+		errCh <- b.errGroup.Wait()
```

Comment:
```suggestion
		err:= <- b.errGroup.Wait()
		select {
		case errC<-err:
		case <-ctx.Done():
		}
```

this avoids leaking memory